ops/model.py

```diff
@@ -60,9 +60,13 @@ def get_relation(self, relation_name, relation_id=None):
         """
         return self.relations._get_unique(relation_name, relation_id)
 
-    def get_binding(self, relation):
-        """Get a network space binding for a Relation object."""
-        return self._bindings.get(relation)
+    def get_binding(self, binding_key):
+        """Get a network space binding.
+
+        binding_key -- a name of a relation or a Relation object. Providing a Relation object will return
+                       a Binding object specific to a relation.
+        """
```

Comment:
I've been trying to rephrase this to be clearer three or four times now, and failing.

The docstring as it stands just adds to the confusion, and unless you know what a "Binding object specific to a relation" is, and what its ... contrapositive? is, then it's not at all clear what you get if you pass in a string.

I get that "bindings" are the juju names for these, and "extra bindings" being bindings that aren't binding anything with anything are part of the problem we're fixing longer-term in juju itself, but this conflation here doesn't seem to help.

Where is this used, such that we want a single method? Wouldn't it make more sense to have a `get_binding` and a `get_binding_by_name`? (or maybe even a `get_extra_binding` and we treat "extra binding" as a magic term that juju gives us that has little connection with the other "binding" thing juju gives us?)

Reply:
This is the case we discussed previously: we're trying to cut down on the required API for it while we don't have a crystal clear understanding of what the new version of that feature will look like, so we don't have to be obsoleting multiple entry points if possible.

Here is some suggested phrasing:
> """Get a network space binding.
> 
> binding_key -- The relation name or instance to obtain bindings for.
>
> If binding_key is a relation name the method returns the default binding for that relation. If a relation instance is provided the method first looks up a more specific binding for that specific relation ID, and if none is found falls back to the default binding for the relation name.
> """

Reply:
> returns the default binding for that relation.

`default binding for a relation` is a new term. Are we OK with introducing it here?

There is an existing term `a default binding for an application` which means that any relation endpoint will be bound to a specified space by default unless a different one is specified.

https://github.com/juju/juju/blob/27d29321e2ab68d60310d136de8daa24d1266c8a/cmd/juju/application/bind.go#L83
https://github.com/juju/juju/blob/27d29321e2ab68d60310d136de8daa24d1266c8a/core/network/space.go#L16-L20 (also named as the "alpha" space in the newer code base)

For example:

metadata:
```
provides:
    db:
        interface: db
    logging:
        interface: logging
    monitoring:
        interface: monitoring
```
bundle.yaml, bindings section of an app:
```
...
applications:
  mycharm:
    # ...
    bindings:
      "": my-default-space # this is a syntax for specifying a default binding
      db: red-space
      # both "monitoring" and "logging" endpoints will be bound to "my-default-space"
```

`get_binding('db').network.<something>` will return network info not tied to a particular relation ID but from the "red-space" (not a space provided in a default space binding for this application).

I would like to avoid overloading the term "default binding" for an app vs a relation because it is already confusing enough.

@chipaca I originally had two methods in one of the previous PRs but we decided to go for a single one for the reasons @niemeyer is outlining.

Reply:
Okay, here is some food for thought:

1. What will the method return when you use a relation name?
2. What will the method return when you use a relation instance and there was a binding to that particular relation instance?
3. What will the method return when you use a relation instance and there is a binding to the relation name but not the relation instance?

What's the proposal for naming things in these scenarios?

Reply:
Based on the definition of an application endpoint I would have "app endpoint bindings" and "relation-specific app endpoint bindings":

https://jaas.ai/docs/concepts-and-terms#heading--endpoint
> An endpoint (or application endpoint) is used to connect to another application’s endpoint in order to form a relation.

My logic is as follows:

* we can only bind app endpoints;
* a relation is formed by _connecting two endpoints_ of different apps and gets an id;
* The lookup of network info is affected by 3 factors: the **current app**'s endpoint bindings, endpoint name, relation id. This means that using a term "relation binding" is not correct because we do not bind relations themselves.

So:

1. app endpoint binding;
2. relation-specific app endpoint binding;
3. app endpoint binding.

Reply:
@dshcherb We seem to be in different wavelengths.  I don't think anyone would understand if we said "we return the app endpoint binding" in any of those cases.

Reply:
https://github.com/canonical/operator/pull/157/commits/cea632893deb9afe259a2dab23d6f1fd12076e0e